operations/kaas/cluster-stack-operator.md

```diff
@@ -0,0 +1,692 @@
+    - **Built-in Predicates**: There are several ready-to-use predicates, like `resource.CreatePredicate`, `resource.UpdatePredicate`, `resource.DeletePredicate`, and `resource.GenericPredicate`. They filter out events based on create, update, delete, or generic actions on resources.
+    - **Custom Predicates**: For more fine-grained control, we can create custom predicates by implementing the `Predicate` interface. This interface defines functions like `Create`, `Update`, `Delete`, and `Generic`, which allow us to specify whether or not to handle respective events.
+
+Using these methods, our controllers get called whenever the events we are interested in (and haven't filtered out) occur, which greatly improves the user experience. As we have dependencies between resources that depend on updates of each other, the builder package makes quick updates of all resources possible. 
+
+### Testing
+
+To ensure the robustness and reliability of our code, it's imperative that we subject it to rigorous testing. Testing controllers, especially those with a reconcile loop initiated by a manager, poses its own set of unique challenges and nuances; it isn't always a straightforward task.
+
+To facilitate this, we'll leverage the capabilities of the envtest package. envtest is an invaluable tool that comes from the Kubernetes controller-runtime project. It is designed to streamline integration tests by furnishing the requisite clientsets, caches, and assorted components necessary for executing a local Kubernetes API server and etcd instances.
+
+What does this mean for us? This functionality paves the way for testing Kubernetes controllers and evaluating their interactions with the Kubernetes API without the prerequisites of a physical cluster. This is a significant advantage as it means we can authenticate the complete functionality of our code, sans the need for an actual cluster. This is applicable both for local testing and within our Continuous Integration (CI) environments.
+
+However, it's essential to be aware that our setup does have multiple external dependencies that need to be addressed to ensure comprehensive testing.
+
+To start with, we require a local kind cluster. This cluster functions as the foundation for our workload during local tests. By utilizing this kind cluster, we are endowed with the ability to emulate the process of applying cluster addons directly to a workload cluster.
+
+Moreover, for our testing regimen, we also have local releases of cluster stacks. These are particularly crucial for unit tests as they negate the need to rely on an external Git repository. In order to check that downloads work from e.g. Github, we add specific dedicated tests to make sure that the download works properly. 
+
+### Repo DevOps: Tools for Efficient Development
+
+#### kubebuilder
```

Comment:
yes kubebuilder might minimize boilerplate code, but this whole paragraph could have been left out...